chain/client/src/test_utils.rs

```diff
@@ -1102,7 +1102,151 @@ pub struct TestEnv {
+        self.runtime_adapters = Some(adapters);
+        self
+    }
+
+    /// Specifies custom network adaptors for each client.
+    ///
+    /// The vector must have the same number of elements as they are clients.
+    /// If that does not hold, `build` method will panic.
+    pub fn network_adapters(mut self, adapters: Vec<Arc<MockNetworkAdapter>>) -> Self {
+        self.network_adapters = Some(adapters);
+        self
+    }
+
+    /// Constructs new `TestEnv` structure.
+    ///
+    /// If no clients were configured (either through count or vector) one
+    /// client is created.  Similarly, if no validator seats were configured,
+    /// one seat is configured.
+    ///
+    /// Panics if `runtime_adapters` or `network_adapters` methods were used and
+    /// the length of the vectors passed to them did not equal number of
+    /// configured clients.
+    pub fn build(self) -> TestEnv {
+        let chain_genesis = self.chain_genesis;
+        let clients = self.clients;
+        let num_clients = clients.len();
+        let validators = self.validators;
+        let num_validators = validators.len();
+        let network_adapters = self
+            .network_adapters
+            .unwrap_or_else(|| (0..num_clients).map(|_| Arc::new(Default::default())).collect());
+        assert!(clients.len() == network_adapters.len());
+        let clients = match self.runtime_adapters {
+            None => clients
+                .into_iter()
+                .zip(network_adapters.iter())
+                .map(|(account_id, network_adapter)| {
+                    setup_client(
+                        create_test_store(),
+                        vec![validators.clone()],
+                        1,
+                        1,
+                        Some(account_id),
+                        false,
+                        network_adapter.clone(),
+                        chain_genesis.clone(),
+                    )
+                })
+                .collect(),
+            Some(runtime_adapters) => {
+                assert!(clients.len() == runtime_adapters.len());
+                clients
+                    .into_iter()
+                    .zip((&network_adapters).iter())
+                    .zip(runtime_adapters.into_iter())
+                    .map(|((account_id, network_adapter), runtime_adapter)| {
+                        setup_client_with_runtime(
+                            u64::try_from(num_validators).unwrap(),
+                            Some(account_id),
+                            false,
+                            network_adapter.clone(),
+                            chain_genesis.clone(),
+                            runtime_adapter,
+                        )
+                    })
+                    .collect()
+            }
+        };
+
+        TestEnv { chain_genesis, validators, network_adapters, clients }
+    }
+
+    fn default_formatter(id: usize) -> std::string::String {
```

Comment:
Just `String`?

Reply:
Yeah. Admittedly this is a vestige of the earlier version of the code. The idea is to let users provide custom formatters (though currently that feature is not available) and in that case it seems that the typical use case will be to call `format!` which returns a `String`. I therefore moved converting the string to `AccountId` inside of the builder.  This could also by anything implementing `TryInto<AccountId>` but that adds more generic types which I didn’t care to define.

Reply:
Ah, wait, you just meant to use `String` i/o `std::string::String`. Will do in a follow up commit.